src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs

```diff
@@ -0,0 +1,165 @@
+            }
+            else
+            {
+                Debug.Assert(info.DiagnosticInfo == null || addDependencies); // It would be strange to drop diagnostics, but record dependencies
+            }
+
+            if (addDependencies)
+            {
+                useSiteInfo.AddDependencies(info);
+            }
+        }
+    }
+
+    internal sealed class BindingDiagnosticBag : BindingDiagnosticBag<AssemblySymbol>
+    {
+        public static readonly BindingDiagnosticBag Discarded = new BindingDiagnosticBag(null, null);
+
+        public BindingDiagnosticBag()
+            : this(usePool: false)
+        { }
+
+        private BindingDiagnosticBag(bool usePool)
+            : base(usePool)
+        { }
+
+        public BindingDiagnosticBag(DiagnosticBag? diagnosticBag)
+            : base(diagnosticBag, dependenciesBag: null)
+        {
+        }
+
+        public BindingDiagnosticBag(DiagnosticBag? diagnosticBag, ICollection<AssemblySymbol>? dependenciesBag)
+            : base(diagnosticBag, dependenciesBag)
+        {
+        }
+
+        internal static BindingDiagnosticBag GetInstance()
+        {
+            return new BindingDiagnosticBag(usePool: true);
+        }
+
+        // PROTOTYPE(UsedAssemblyReferences): Remove this overload once conversion operators are removed from the base type.
+        internal void AddRange(BindingDiagnosticBag other)
+        {
+            AddRange((BindingDiagnosticBag<AssemblySymbol>)other);
+        }
+
+        internal void AddDependencies(Symbol? symbol)
+        {
+            if (symbol is object && DependenciesBag is object)
+            {
+                AddDependencies(symbol.GetUseSiteInfo());
+            }
+        }
+
+        internal bool ReportUseSite(Symbol? symbol, SyntaxNode node)
+        {
+            return ReportUseSite(symbol, node.Location);
+        }
+
+        internal bool ReportUseSite(Symbol? symbol, SyntaxToken token)
+        {
+            return ReportUseSite(symbol, token.GetLocation());
+        }
+
+        internal bool ReportUseSite(Symbol? symbol, Location location)
+        {
+            if (symbol is object)
+            {
+                return Add(symbol.GetUseSiteInfo(), location);
+            }
+
+            return false;
+        }
+
+        internal void AddAssembliesUsedByNamespaceReference(NamespaceSymbol ns)
+        {
+            if (DependenciesBag is null)
+            {
+                return;
+            }
+
+            addAssembliesUsedByNamespaceReference(ns);
+
+            void addAssembliesUsedByNamespaceReference(NamespaceSymbol ns)
+            {
+                // Treat all assemblies contributing to this namespace symbol as used
+                if (ns.Extent.Kind == NamespaceKind.Compilation)
+                {
+                    foreach (var constituent in ns.ConstituentNamespaces)
+                    {
+                        AddAssembliesUsedByNamespaceReference(constituent);
```

Comment:
>AddAssembliesUsedByNamespaceReference [](start = 24, length = 37)

This should call the containing local function instead of it's enclosing function. There is no correctness problem here though, just a missing optimization.

Reply:
Adjusted in [https://github.com/dotnet/roslyn/pull/40640](https://github.com/dotnet/roslyn/pull/40640)

---
In reply to: [361555849](https://github.com/dotnet/roslyn/pull/40560#discussion_r361555849) [](ancestors = 361555849)